### PR TITLE
MemCache - a minion data cache booster.

### LINKED ISSUE #39275: Cache backend gets hit a *lot*
### Description of Issue/Question
It turns out that the cache backend fech() function gets hit a *LOT*. When I put consul as a cache backend in an env with ~200 minions, salt became unusable due to the consul process getting overloaded.
When I put a print statement in the fetch() function, it seems it gets called 400+ times *per minion*, making the consul backend essentially unusable.

### Setup
set `cache: consul` in the master config and count the number you hit fetch() in salt/cache/consul.py

### Versions Report
Salt Version:
           Salt: 2016.11.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-100-generic
         system: Linux
        version: Ubuntu 14.04 trusty

### PR DESCRIPTION
**Pull request by DmitryKuzmenko**
### What does this PR do?
Introduce a minion data cache booster `memcache`.

### What issues does this PR fix or reference?
Fixes #39275

### Tests written?
Yes

### CODE REVIEW

**Comment by DmitryKuzmenko** [comment]
I'll add the docs later in further commits.

**Review by cachedout — APPROVED** [review]

File: salt/config/__init__.py [hunk 1]
```diff
@@ -919,6 +919,14 @@ def _gather_buffer_space():
 
     # Minion data cache driver (one of satl.cache.* modules)
     'cache': str,
+    # Enables a fast in-memory cache booster and sets the expiration time.
+    'memcache_expire_seconds': int,
```
Comment (mirceaulinic) [inline salt/config/__init__.py]: I was just wondering: how is this going to influence the caching etc? I would be worried if this requires to execute `refresh_pillar` more often, for example.

Comment (DmitryKuzmenko) [inline salt/config/__init__.py]: It's just about keeping some of `minion_data_cache` data in-memory to don't fetch them from the cache backend again and again. No any additional cache logic is here. With and without it the business logic has kept the same.

Comment (mirceaulinic) [inline salt/config/__init__.py]: Oh, right, it's about retrieving from the cache database, not the caching itself.

Thanks @DmitryKuzmenko

**Pull request merged**